File: lossdiafaq/services/calculator/ees.py
import math
import random
import sys

import lossdiafaq.static as static


__all__ = ["EESCalculator"]


class EESArgumentError(Exception):
    """Error with EES arguments"""


class EESCalculator:
    """Simulates using EES from start to end
    
    Originally written by Optimist#6937
    Modified slightly by jacοb#0828
    
    Attributes
    ----------
    start : :class:`int`

    end : :class:`int`

    delta : :class:`int`

    avg_attempts : :class:`float`
        The average amount of attempts it takes to go from start to end
    
    avg_sfprots_used : :class:`float`
        The average amount of starforce protects it takes to go from start to end

    avg_meso_used : :class:`float`
        The average amount of meso it takes to go from start to end

    avg_vp_used : :class:`float`
        The average amount of VP it takes to go from start to end

    avg_credits_used : :class:`float`
        The average amount of credits it takes to go from start to end

    min_attempts : :class:`float`
        The minimum amount of attempts it took to go from start to end
    
    min_sfprots_used : :class:`float`
        The minimum amount of starforce protects it took to go from start to end

    min_meso_used : :class:`float`
        The minimum amount of meso it took to go from start to end

    min_vp_used : :class:`float`
        The minimum amount of VP it took to go from start to end

    min_credits_used : :class:`float`
        The minimum amount of credits it took to go from start to end

    max_attempts : :class:`float`
        The maximum amount of attempts it took to go from start to end
    
    max_sfprots_used : :class:`float`
        The maximum amount of starforce protects it took to go from start to end

    max_meso_used : :class:`float`
        The maximum amount of meso it took to go from start to end

    max_vp_used : :class:`float`
        The maximum amount of VP it took to go from start to end

    max_credits_used : :class:`float`
        The maximum amount of credits it took to go from start to end
    """
    def __init__(self, start: int, end: int, delta: int, *, aees=False) -> None:
        self.start = start
        self.end = end
        self.delta = delta

        self.avg_attempts = 0.
        self.avg_sfprots_used = 0.
        self.avg_meso_used = 0.
        self.avg_vp_used = 0.
        self.avg_credits_used = 0.
        self.avg_cogs_used = 0.
        self.avg_ees_used = 0.

        self.min_attempts = sys.maxsize
        self.min_sfprots_used = sys.maxsize
        self.min_meso_used = sys.maxsize
        self.min_vp_used = sys.maxsize
        self.min_credits_used = sys.maxsize
        self.min_cogs_used = sys.maxsize
        self.min_ees_used = sys.maxsize

        self.max_attempts = -1
        self.max_sfprots_used = -1
        self.max_meso_used = -1
        self.max_vp_used = -1
        self.max_credits_used = -1
        self.max_cogs_used = -1
        self.max_ees_used = -1

        self.base_rate = static.LOSSDIA_AEES_BASE_RATE if aees else static.LOSSDIA_EES_BASE_RATE
        self.min_rate = static.LOSSDIA_AEES_MIN_RATE if aees else static.LOSSDIA_EES_MIN_RATE

    def simulate(self):
        """Does the EES simulation"""
        if self.start < 0 or self.start > 14:
            raise EESArgumentError("Start must be within 0-14")
            
        if self.end < 1 or self.end > 15:
            raise EESArgumentError("End must be within 1-15")
            
        if self.delta < 0 or self.delta > 4:
            raise EESArgumentError("Protect delta must be within 0-4")

        samples = static.LOSSDIA_EES_SAMPLES
        meso_cost = static.LOSSDIA_EES_MESO_COST
        vp_cost = static.LOSSDIA_EES_SFPROT_VP_COST
        dp_cost = static.LOSSDIA_EES_SFPROT_CREDITS_COST
        cogs_cost = static.LOSSDIA_AEES_COG_COST
        ees_cost = static.LOSSDIA_AEES_EES_COST

        for _ in range(samples):
            attempts, sfprots_used = self._attempt_ees()
            meso_used = attempts * meso_cost
            vp_used = sfprots_used * vp_cost
            dp_used = sfprots_used * dp_cost
            cogs_used = attempts * cogs_cost
            ees_used = attempts * ees_cost

            self.avg_attempts += attempts / samples
            self.avg_sfprots_used += sfprots_used / samples
            self.avg_meso_used += meso_used / samples
            self.avg_credits_used += dp_used / samples
            self.avg_vp_used += vp_used / samples
            self.avg_cogs_used += cogs_used / samples
            self.avg_ees_used += ees_used / samples

            if attempts < self.min_attempts:
                self.min_attempts = attempts
            elif attempts > self.max_attempts:
                self.max_attempts = attempts

            if meso_used < self.min_meso_used:
                self.min_meso_used = meso_used
            elif meso_used > self.max_meso_used:
                self.max_meso_used = meso_used

            if vp_used < self.min_vp_used:
                self.min_vp_used = vp_used
            elif vp_used > self.max_vp_used:
                self.max_vp_used = vp_used

            if dp_used < self.min_credits_used:
                self.min_credits_used = dp_used
            elif dp_used > self.max_credits_used:
                self.max_credits_used = dp_used

            if sfprots_used < self.min_sfprots_used:
                self.min_sfprots_used = sfprots_used
            elif sfprots_used > self.max_sfprots_used:
                self.max_sfprots_used = sfprots_used

            if cogs_used < self.min_cogs_used:
                self.min_cogs_used = cogs_used
            elif cogs_used > self.max_cogs_used:
                self.max_cogs_used = cogs_used

            if ees_used < self.min_ees_used:
                self.min_ees_used = ees_used
            elif ees_used > self.max_ees_used:
                self.max_ees_used = ees_used

        return

    def _enhance(self, level: int) -> tuple[int, bool]:
        rate = math.ceil(max(self.base_rate - static.LOSSDIA_EES_REDUCTION_RATE * level, self.min_rate) * 100)
        random_result = random.randint(0, 100)
        ees_result = random_result <= rate

        sfprot_used = (level % static.LOSSDIA_EES_MIN_SFPROT_LEVEL) >= (static.LOSSDIA_EES_MIN_SFPROT_LEVEL - self.delta) and level > static.LOSSDIA_EES_MIN_SFPROT_LEVEL

        if not ees_result and level % static.LOSSDIA_EES_MIN_SFPROT_LEVEL == 0:
            # level won't move at safe points (5, 10, 15) if EES fails
            return level, sfprot_used
        elif not ees_result:
            return level - 1 + int(sfprot_used), sfprot_used
        else:
            return level + 1, sfprot_used

    def _attempt_ees(self) -> tuple[int, int]:
        current_level = self.start
        attempts = 0
        sfprots_used = 0
        
        while current_level < self.end:
            current_level, sfprot_used = self._enhance(current_level)
            if sfprot_used:
                sfprots_used += 1
            
            attempts += 1

        return attempts, sfprots_used
    

File: server/internal/calculator/ees.go
/*
Credit: Optimist#6397
*/
package calculator

import (
	"errors"
	"math"
	"math/rand"

	"github.com/jaczerob/server/internal/utils/static"
)

type EESCalculator struct {
	Start, End, Delta float64
	baseRate, minRate float64

	AvgAttempts, AvgSFProtsUsed, AvgMesoUsed, AvgVPUsed, AvgCreditsUsed, AvgCOGsUsed, AvgEESUsed float64
	MinAttempts, MinSFProtsUsed, MinMesoUsed, MinVPUsed, MinCreditsUsed, MinCOGsUsed, MinEESUsed float64
	MaxAttempts, MaxSFProtsUsed, MaxMesoUsed, MaxVPUsed, MaxCreditsUsed, MaxCOGsUsed, MaxEESUsed float64
}

func NewEESCalculator(start, end, delta, baseRate, minRate float64) (e *EESCalculator, err error) {
	if start < 0 || start > 14 {
		return nil, errors.New("Start must be within 0-14")
	}

	if end < 1 || end > 15 {
		return nil, errors.New("End must be within 1-15")
	}

	if delta < 0 || delta > 4 {
		return nil, errors.New("Delta must be within 0-4")
	}

	e = &EESCalculator{
		Start:    start,
		End:      end,
		Delta:    delta,
		baseRate: baseRate,
		minRate:  minRate,

		MinAttempts:    math.MaxFloat64,
		MinSFProtsUsed: math.MaxFloat64,
		MinMesoUsed:    math.MaxFloat64,
		MinVPUsed:      math.MaxFloat64,
		MinCreditsUsed: math.MaxFloat64,
		MinCOGsUsed:    math.MaxFloat64,
		MinEESUsed:     math.MaxFloat64,
		MaxAttempts:    -1.,
		MaxSFProtsUsed: -1.,
		MaxMesoUsed:    -1.,
		MaxVPUsed:      -1.,
		MaxCreditsUsed: -1.,
		MaxCOGsUsed:    -1.,
		MaxEESUsed:     -1.,
	}

	return
}

func (e *EESCalculator) Calculate() {
	samples := static.WindiaEESSamples

	for i := 0.; i < samples; i++ {
		attempts, sfProtsUsed := e.attemptEES()
		mesoUsed := attempts * static.WindiaEESMesoCost
		cogsUsed := attempts * static.WindiaAEESCOGCost
		eesUsed := attempts * static.WindiaAEESEESCost
		vpUsed := sfProtsUsed * static.WindiaEESSFProtVPCost
		creditsUsed := sfProtsUsed * static.WindiaEESSFProtCreditsCost

		e.AvgAttempts += attempts / samples
		e.AvgSFProtsUsed += sfProtsUsed / samples
		e.AvgMesoUsed += mesoUsed / samples
		e.AvgVPUsed += vpUsed / samples
		e.AvgCreditsUsed += creditsUsed / samples
		e.AvgCOGsUsed += cogsUsed / samples
		e.AvgEESUsed += eesUsed / samples

		if attempts < e.MinAttempts {
			e.MinAttempts = attempts
		} else if attempts > e.MaxAttempts {
			e.MaxAttempts = attempts
		}

		if mesoUsed < e.MinMesoUsed {
			e.MinMesoUsed = mesoUsed
		} else if mesoUsed > e.MaxMesoUsed {
			e.MaxMesoUsed = mesoUsed
		}

		if vpUsed < e.MinVPUsed {
			e.MinVPUsed = vpUsed
		} else if vpUsed > e.MaxVPUsed {
			e.MaxVPUsed = vpUsed
		}

		if creditsUsed < e.MinCreditsUsed {
			e.MinCreditsUsed = creditsUsed
		} else if creditsUsed > e.MaxCreditsUsed {
			e.MaxCreditsUsed = creditsUsed
		}

		if sfProtsUsed < e.MinSFProtsUsed {
			e.MinSFProtsUsed = sfProtsUsed
		} else if sfProtsUsed > e.MaxSFProtsUsed {
			e.MaxSFProtsUsed = sfProtsUsed
		}

		if cogsUsed < e.MinCOGsUsed {
			e.MinCOGsUsed = cogsUsed
		} else if cogsUsed > e.MaxCOGsUsed {
			e.MaxCOGsUsed = cogsUsed
		}

		if eesUsed < e.MinEESUsed {
			e.MinEESUsed = eesUsed
		} else if eesUsed > e.MaxEESUsed {
			e.MaxEESUsed = eesUsed
		}
	}
}

func (e *EESCalculator) enhance(level float64) (newLevel float64, sfProtUsed bool) {
	rate := math.Ceil(math.Max(e.baseRate-static.WindiaEESReductionRate*level, e.minRate) * 100)
	eesResult := float64(rand.Intn(101)) <= rate

	levelsFromSafepoint := math.Mod(level, 5.)
	sfProtUsed = levelsFromSafepoint >= (5.-e.Delta) && level > 5

	if eesResult {
		newLevel = level + 1
	} else if !eesResult && !sfProtUsed && levelsFromSafepoint != 0 {
		newLevel = level - 1
	} else {
		newLevel = level
	}

	return
}

func (e *EESCalculator) attemptEES() (attempts, sfProtsUsed float64) {
	level := e.Start
	for level < e.End {
		newLevel, sfProtUsed := e.enhance(level)
		attempts += 1.

		if sfProtUsed {
			sfProtsUsed += 1
		}

		level = newLevel
	}

	return
}
